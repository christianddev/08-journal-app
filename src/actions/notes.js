
import Swal from 'sweetalert2';
import {db} from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';


export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
})

export const addNewNote = (id, note) => ({
  type: types.notesAddNote,
  payload:{
    id,
    ...note
  }
})

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const {auth: {uid}} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    try {
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      dispatch(activeNote(doc.id, newNote));
      dispatch(addNewNote(doc.id, newNote));
    } catch (error) {
      console.log('error', error)
    }
  }
}

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {auth: {uid}} = getState();
    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    try {
      await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
      dispatch(refreshNote(note.id, noteToFirestore));
      Swal.fire({
        title: 'Saved',
        text: note.title,
        icon: 'success'
      });
    } catch (error) {
      Swal.fire({
        title: error?.code,
        text: error?.message,
        icon: 'error'
      });
      console.error('error startSaveNote: ', error);
    }
  }
}

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id, note: {
      id,
      ...note
    }
  }
})

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const {notes: {active: note}} = getState();
    Swal.fire({
      title: 'Uploading',
      text: 'Please wait',
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      }
    });
    const fileUrl = await fileUpload(file);

    Swal.close();

    note.url = fileUrl;
    dispatch(startSaveNote(note));

  }
}

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: {
    id
  }
})

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    try {
      await db.doc(`${uid}/journal/notes/${id}`).delete()
      dispatch(deleteNote(id));
      Swal.fire({
        title: 'Deleted',
        text: 'Picture',
        icon: 'success'
      });
    } catch (error) {
      Swal.fire({
        title: error?.code,
        text: error?.message,
        icon: 'error'
      });
      console.error('error startDeleting: ', error);
    }
  }
}

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
});

