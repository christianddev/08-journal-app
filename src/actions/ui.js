import { types } from "../types/types"

export const serError = (error) => ({
  type: types.uiSetError,
  payload: error
})

export const removeError = () => ({
  type: types.uiRemoveSetError
})

export const startLoading= () => ({
  type: types.uiStartLoading
})

export const finishLoading = () => ({
  type: types.uiFinishLoading
})