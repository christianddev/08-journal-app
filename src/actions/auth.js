import Swal from 'sweetalert2';
import { types } from '../types/types';
import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const login = (uid, displayName) => (
  {
    type: types.login,
    payload: {
      uid,
      displayName
    }
  }
)

export const logout = () => (
  {
    type: types.logout
  }
);

export const startLoginWithEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({user}) => {
        dispatch(finishLoading());
        dispatch(
          login(user.uid, user.displayName)
        );
      } )
      .catch((error) => {
        Swal.fire({
          title: error?.code,
          text: error?.message,
          icon: 'error'
        });
        console.error('error startLoginWithEmailPassword: ', error);
        dispatch(finishLoading());
      });
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return ( dispatch) => {
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then( async({user}) => {
        await user.updateProfile({displayName: name})
        dispatch(
          login(user.uid, user.displayName)
        );
      })
      .catch((error) => {
        Swal.fire({
          title: error?.code,
          text: error?.message,
          icon: 'error'
        });
        console.error('error startRegisterWithEmailPasswordName: ', error)
      }
        );
  }

}

export const startLoginGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({user}) => {
        dispatch(login(user.uid, user.displayName));
      })
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch( noteLogout() );
  }
}