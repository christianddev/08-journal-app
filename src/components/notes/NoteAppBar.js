import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { startSaveNote, startUploading } from '../../actions/notes';

const NotesAppBar = () => {
    const dispatch = useDispatch();
    const {notes: {active: note}} = useSelector(state => state);

    const noteDate = moment(note.date);
    const handleSave = () => {
        dispatch(startSaveNote(note));
    }

    const handlePictureUpload = (e) => {
        document.querySelector('#fileSelector').click();
    }

    const handleInputChange = (e) => {
        const file = e.target.files[0];
        if (file) {
            dispatch(startUploading(file));
        }
    }

    startSaveNote()
    return (
        <div className="notes__appbar">
            <span>{noteDate.format('Do MMMM YYYY')}</span>
            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{display: 'none'}}
                onChange={handleInputChange}
            />

            <div>
                <button
                    className="btn"
                    onClick={handlePictureUpload}
                >
                    Picture
                </button>

                <button
                    className="btn"
                    onClick={handleSave}
                    >
                    Save
                </button>
            </div>
        </div>
    )
}

export default NotesAppBar;