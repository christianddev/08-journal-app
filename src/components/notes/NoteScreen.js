import React from 'react';
import NoteAppBar from './NoteAppBar';

const NoteScreen = () => {
    return (
        <div className="notes__main-content">
            <NoteAppBar/>

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                 />
                 <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                 ></textarea>
                 <div className="notes__image">
                    <img 
                        src="https://image.freepik.com/free-vector/watercolor-style-natural-landscape-background_23-2148218859.jpg" 
                        alt="note_image"
                        />

                 </div>
            </div>
        </div>
    )
}

export default NoteScreen;
