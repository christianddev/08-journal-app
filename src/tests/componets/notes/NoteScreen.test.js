import React from 'react';
import { mount } from "enzyme"
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { activeNote } from '../../../actions/notes';
import NoteScreen from '../../../components/notes/NoteScreen';

jest.mock('../../../actions/notes', () => ({
  activeNote: jest.fn(),
  startDeleting: jest.fn(),
  startSaveNote: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
  auth:{},
  ui:{
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id: 'abc',
      title: 'titleXD',
      body: 'bodyXD'
    },
    notes:  []
  }
}
let store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount (
  <Provider store={store}>
    <MemoryRouter>
      <NoteScreen />
    </MemoryRouter>
  </Provider>
)

describe('NoteScreen.js', () => {
  test('should show', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should call activeNote', () => {
    const event = {
      target: {
        name: 'title',
        value: 'titleDemo'
      }
    }
    wrapper.find('input[name="title"]').simulate("change", event);
    expect(activeNote).toHaveBeenLastCalledWith(
      'abc',
      {
        id: 'abc',
        title: 'titleDemo',
        body: 'bodyXD'
      }
    )
  })


})
