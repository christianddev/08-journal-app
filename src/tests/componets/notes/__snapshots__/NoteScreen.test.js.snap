// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteScreen.js should show 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      23rd October 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="titleXD"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="bodyXD"
    />
  </div>
  <button
    className="btn btn-sanger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
